File: md2d.zig
//
// The MIT License
//
// Copyright 2022 David D. Clark.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

//
// A molecular dynamics simulation of a two-dimensional Lennard-Jones fluid.
//
// Author: david
//

const std = @import("std");
const builtin = @import("builtin");
const math = std.math;
const cTime = @cImport(@cInclude("time.h"));

// Simulation parameters.
const num_particles: i32 = 500;
const force_cutoff: f64 = 3.0;
const force_cutoff_2: f64 = force_cutoff * force_cutoff;
const p_eat_cutoff: f64 = 4.0 * (math.pow(f64, force_cutoff, -12.0) - math.pow(f64, force_cutoff, -6.0));
const target_temperature: f64 = 0.42;
const time_step: f64 = 0.020;
const dt_over_2 = 0.5 * time_step;
const dt_squared_over_2 = 0.5 * time_step * time_step;
const box_width: f64 = 100.0;
const box_width_minus_half: f64 = box_width - 0.5;
const box_height: f64 = 100.0;
const box_height_minus_half: f64 = box_height - 0.5;
const wall_stiffness: f64 = 50.0;
const equilibration_time: f64 = 200.0;
const steps_between_equil_rescaling: i32 = 10;
const production_time: f64 = 200.0;
const steps_between_prod_rescaling: i32 = 100;
const steps_per_printout: i32 = 50;

// Other globals.
var x: [num_particles]f64 = std.mem.zeroes([num_particles]f64);
var y: [num_particles]f64 = std.mem.zeroes([num_particles]f64);
var vx: [num_particles]f64 = std.mem.zeroes([num_particles]f64);
var vy: [num_particles]f64 = std.mem.zeroes([num_particles]f64);
var ax: [num_particles]f64 = std.mem.zeroes([num_particles]f64);
var ay: [num_particles]f64 = std.mem.zeroes([num_particles]f64);

var t: f64 = undefined; // Simulation time
var steps_accomplished: i32 = 0;

var v_sum_x: f64 = 0.0;
var v_sum_y: f64 = 0.0;
var k_e: f64 = 0.0;
var p_e: f64 = 0.0;
var current_t: f64 = 0.0;
var total_t: f64 = 0.0;
var average_t: f64 = 0.0;
var pressure: f64 = 0.0;
var total_p: f64 = 0.0;
var average_p: f64 = 0.0;
var sample_count: i32 = 0;

pub fn main() void {
    print("md2d - MD simulation of a 2D argon gas Lennard-Jones system.\n", .{});
    print("This version is written in Zig and running version: {}\n", .{builtin.zig_version});

    printConfig();

    initializeParticles();
    print("\nEquilibration\n", .{});
    printPropertiesHeader();
    printProperties();
    var start: cTime.clock_t = cTime.clock();
    while (t < equilibration_time) {
        singleStep();
        if (steps_accomplished == 0) {
            printProperties();
        }
        steps_accomplished += 1;
        t += time_step;
        if (@mod(steps_accomplished, steps_per_printout) == 0) {
            computeProperties();
            printProperties();
        }
        if (@mod(steps_accomplished, steps_between_equil_rescaling) == 0) {
            rescaleVelocities();
        }
    }

    resetMeasurements();
    rescaleVelocities();
    print("\nProduction\n", .{});
    printPropertiesHeader();
    while (t < (equilibration_time + production_time)) {
        singleStep();
        steps_accomplished += 1;
        t += time_step;
        if (@mod(steps_accomplished, steps_per_printout) == 0) {
            computeProperties();
            printProperties();
        }
        if (@mod(steps_accomplished, steps_between_prod_rescaling) == 0) {
            rescaleVelocities();
        }
    }

    var finish: cTime.clock_t = cTime.clock();
    var elapsed: f64 = @intToFloat(f64, (finish - start)) / @intToFloat(f64, cTime.CLOCKS_PER_SEC);
    print("Elapsed time: {d:.3} seconds, {d:.3} steps per second.\n", .{ elapsed, @intToFloat(f64, steps_accomplished) / elapsed });
    std.os.exit(0);
}

/// Execute one time step using the Verlet algorithm.
fn singleStep() void {
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        // Update positions.
        x[i] += (vx[i] * time_step) + (ax[i] * dt_squared_over_2);
        y[i] += (vy[i] * time_step) + (ay[i] * dt_squared_over_2);
        // "Half update" the velocities.
        vx[i] += ax[i] * dt_over_2;
        vy[i] += ay[i] * dt_over_2;
    }
    computeAccelerations();
    // Complete updates to velocities with new accelerations.
    i = 0;
    while (i < num_particles) : (i += 1) {
        vx[i] += ax[i] * dt_over_2;
        vy[i] += ay[i] * dt_over_2;
    }
}

// Compute accelerations of the molecules from their current positions using
// the Lennard-Jones potential.
fn computeAccelerations() void {
    var dx: f64 = undefined;
    var dy: f64 = undefined;
    var dx2: f64 = undefined;
    var dy2: f64 = undefined;
    var r_squared: f64 = undefined;
    var r_squared_inv: f64 = undefined;
    var attract: f64 = undefined;
    var repel: f64 = undefined;
    var f_over_r: f64 = undefined;
    var fx: f64 = undefined;
    var fy: f64 = undefined;
    var wall_force: f64 = 0.0;

    p_e = 0.0;

    // First, check for bouncing against the walls.
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        if (x[i] < 0.5) {
            ax[i] = wall_stiffness * (0.5 - x[i]);
            wall_force += ax[i];
            p_e += 0.5 * wall_stiffness * (0.5 - x[i]) * (0.5 - x[i]);
        } else if (x[i] > box_width_minus_half) {
            ax[i] = wall_stiffness * (box_width_minus_half - x[i]);
            wall_force -= ax[i];
            p_e += 0.5 * wall_stiffness * (box_width_minus_half - x[i]) *
                (box_width_minus_half - x[i]);
        } else {
            ax[i] = 0.0;
        }

        if (y[i] < 0.5) {
            ay[i] = wall_stiffness * (0.5 - y[i]);
            wall_force += ay[i];
            p_e += 0.5 * wall_stiffness * (0.5 - y[i]) * (0.5 - y[i]);
        } else if (y[i] > box_height_minus_half) {
            ay[i] = wall_stiffness * (box_height_minus_half - y[i]);
            wall_force -= ay[i];
            p_e += 0.5 * wall_stiffness * (box_height_minus_half - y[i]) *
                (box_height_minus_half - y[i]);
        } else {
            ay[i] = 0.0;
        }
    }

    pressure = wall_force / (4.0 * box_width);

    // Next, compute interactions using the Lennard-Jones potential.
    i = 0;
    var j: usize = 0;
    while (i < num_particles) : (i += 1) {
        j = 0;
        while (j < i) : (j += 1) {
            dx = x[i] - x[j];
            dx2 = dx * dx;
            // Make sure the pair are close enough to bother.
            if (dx2 < force_cutoff_2) {
                dy = y[i] - y[j];
                dy2 = dy * dy;
                if (dy2 < force_cutoff_2) {
                    r_squared = dx2 + dy2;
                    if (r_squared < force_cutoff_2) {
                        r_squared_inv = 1.0 / r_squared;
                        attract = r_squared_inv * r_squared_inv * r_squared_inv;
                        repel = attract * attract;
                        p_e += (4.0 * (repel - attract)) - p_eat_cutoff;
                        f_over_r = 24.0 * ((2.0 * repel) - attract) * r_squared_inv;
                        fx = f_over_r * dx;
                        fy = f_over_r * dy;
                        // Add the force on to i's acceleration.
                        ax[i] += fx;
                        ay[i] += fy;
                        ax[j] -= fx; // Newton's 3rd law.
                        ay[j] -= fy;
                    }
                }
            }
        }
    }
}

/// Reset accumulators and counters used for some measurements.
fn resetMeasurements() void {
    total_t = 0.0;
    total_p = 0.0;
    sample_count = 0;
}

/// Calculate the instantaneous kinetic energy.
fn calculateKineticEnergy() f64 {
    var ek: f64 = 0.0;
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        ek += 0.5 * (vx[i] * vx[i] + vy[i] * vy[i]);
    }
    return ek;
}

/// Compute accumulated property values from sampled values.
fn computeProperties() void {
    sample_count += 1;
    k_e = calculateKineticEnergy();
    current_t = k_e / @intToFloat(f64, num_particles);
    total_t += current_t;
    average_t = total_t / @intToFloat(f64, sample_count);
    total_p += pressure;
    average_p = total_p / @intToFloat(f64, sample_count);
}

/// Return a small random number uniformly distributed around 0
/// with a maximum magnitude of epsilon.
fn nudge(epsilon: f64) f64 {
    return (algo647Uniform() - 0.5) * epsilon;
}

/// Place the particles in empty locations of a grid on the simulation
/// box starting from the lower left. Applies a small amount of "jitter"
/// to break up the regularity of the layout a little. Does not check for
/// overflow of the grid (too many particle to fit).
fn placeParticles() void {
    const spacing: f64 = 1.3; // Minimum space between particle centers.
    const half_spacing: f64 = spacing / 2.0;
    const bwmhs: f64 = box_width - half_spacing;
    const jitter: f64 = 0.01; // Random offset to break up regularity.
    var x_pos: f64 = half_spacing;
    var y_pos: f64 = box_height - half_spacing;

    var idx: usize = 0;
    while (idx < num_particles) : (idx += 1) {
        x[idx] = x_pos + nudge(jitter);
        y[idx] = y_pos + nudge(jitter);
        x_pos += spacing;
        if (x_pos > bwmhs) {
            x_pos = half_spacing;
            y_pos -= spacing;
        }
    }
}

/// Rescale the velocities consistent with the target temperature.
fn rescaleVelocities() void {
    var scaling_factor: f64 = undefined;
    var velocity_squared_sum: f64 = 0.0;
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        velocity_squared_sum += vx[i] * vx[i] + vy[i] * vy[i];
    }
    scaling_factor = 2.0 * @intToFloat(f64, num_particles) * target_temperature / velocity_squared_sum;
    scaling_factor = math.sqrt(scaling_factor);
    i = 0;
    while (i < num_particles) : (i += 1) {
        vx[i] *= scaling_factor;
        vy[i] *= scaling_factor;
    }
}

/// Remove any net momentum from the system of particles.
fn removeDrift() void {
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        v_sum_x += vx[i];
        v_sum_y += vy[i];
    }
    i = 0;
    while (i < num_particles) : (i += 1) {
        vx[i] -= v_sum_x / @intToFloat(f64, num_particles);
        vy[i] -= v_sum_y / @intToFloat(f64, num_particles);
    }
}

/// Initialize particle positions, scale velocities to target temperature,
/// and remove systematic drift.
fn initializeParticles() void {
    var i: usize = 0;
    while (i < num_particles) : (i += 1) {
        x[i] = 0.0;
        y[i] = 0.0;
        vx[i] = gaussianDeviateMarsaglia() - 0.5;
        vy[i] = gaussianDeviateMarsaglia() - 0.5;
        ax[i] = 0.0;
        ay[i] = 0.0;
    }
    placeParticles();
    rescaleVelocities();
    removeDrift();
}

/// Print to stdout, unbuffered, and silently returning on failure.
/// This use of standard output is not really thread safe.
fn print(comptime fmt: []const u8, args: anytype) void {
    const stdout = std.io.getStdOut().writer();
    nosuspend stdout.print(fmt, args) catch return;
}

/// Print out the simulation configuration.
fn printConfig() void {
    print("\nSimulation Configuration:\n", .{});
    print("   Number of particles        : {d}\n", .{num_particles});
    print("   Cutoff radius              : {d}\n", .{force_cutoff});
    print("   Target temperature         : {d}\n", .{target_temperature});
    print("   Integration step size      : {d}\n", .{time_step});
    print("   Width of simulation area   : {d}\n", .{box_width});
    print("   Height of simulation area  : {d}\n", .{box_height});
    print("   Area of simulation         : {d}\n", .{box_width * box_height});
    print("   Wall stiffness             : {d}\n", .{wall_stiffness});
    print("   Equilibration time         : {d}\n", .{equilibration_time});
    print("      Steps between rescaling : {d}\n", .{steps_between_equil_rescaling});
    print("   Production time            : {d}\n", .{production_time});
    print("      Steps between rescaling : {d}\n", .{steps_between_prod_rescaling});
    print("   Steps per print out        : {d}\n", .{steps_per_printout});
}

fn printPropertiesHeader() void {
    print("Time      Temp.   Pressure Tot. E.   Kin. E.   Pot. E.   Steps\n", .{});
}

fn printProperties() void {
    print("{d:>7.3},  {d:>5.3},  {d:>6.4},  {d:>7.2},  {d:>6.2},  {d:>8.2},  {d}\n", .{ t, average_t, average_p, (k_e + p_e), k_e, p_e, steps_accomplished });
}

/// Print some pseudo random numbers. Just for debugging purposes.
fn printRandomNumbers() void {
    print("\nRandom numbers from a uniform distribution:\n", .{});
    var idx: i32 = 0;
    while (idx < 10) : (idx += 1) {
        var rnu: f64 = algo647Uniform();
        print("Uniform #{d} {d}\n", .{ idx, rnu });
    }
    print("\nRandom numbers from a normal distribution:\n", .{});
    idx = 0;
    while (idx < 10) : (idx += 1) {
        var rnn: f64 = gaussianDeviateMarsaglia();
        print("Normal #{d} {d}\n", .{ idx, rnn });
    }
}

// The algo647Uniform function returns a random uniform deviate in
// the range [0, 1) calculated using the ACM algorithm 647 as
// recorded at: http://calgo.acm.org.
// The algorithm is from the paper "ALGORITHM 647: Implementation and
// relative Efficiency of Quasirandom Sequence Generators" by Bennet L. Fox,
// ACM Transactions on Mathematical Software (TOMS), Volume 12, No 4,
// Dec 1986, pp. 362-376.
//
// I have never seen the actual paper - it's behind a paywall - but the
// FORTRAN version of the algorithm is easy enough to understand.

var glbl_seed: i32 = 12345;
const INT32_MAX: i32 = math.maxInt(i32);

/// Return a random uniform deviate in the range 0 <= x < 1.0 as an f64.
fn algo647Uniform() f64 {
    var k: i32 = @divFloor(glbl_seed, 127773);
    // var k: i32 = floor(glbl_seed / 127773);
    var partial: i32 = 16807 * (glbl_seed - k * 127773) - k * 2836;
    if (partial < 0) {
        glbl_seed = partial + INT32_MAX;
    } else {
        glbl_seed = partial;
    }

    return @intToFloat(f64, glbl_seed) * 4.656612875e-10;
}

// The following is an implementation of the Marsaglia polar
// method for generating random standard normal deviates. See
// https://en.wikipedia.org/wiki/Marsaglia_polar_method.
/// Return a random standard normal deviate as an f64.
fn gaussianDeviateMarsaglia() f64 {
    const has_spare = struct {
        var static: bool = false;
    };
    const spare = struct {
        var static: f64 = 0.0;
    };
    var fac: f64 = undefined;
    var rsq: f64 = undefined;
    var r1: f64 = undefined;
    var r2: f64 = undefined;

    if (has_spare.static) {
        has_spare.static = false;
        return spare.static;
    } else {
        while (true) {
            r1 = (2.0 * algo647Uniform()) - 1.0;
            r2 = (2.0 * algo647Uniform()) - 1.0;
            rsq = (r1 * r1) + (r2 * r2);
            if (!((rsq >= 1.0) or (rsq == 0.0))) break;
        }
        fac = math.sqrt(-2.0 * @log(rsq) / rsq);
        spare.static = r1 * fac;
        has_spare.static = true;
        return r2 * fac;
    }
}

File: md2d.jl
#
# The MIT License
#
# Copyright 2022 David D. Clark.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

#
# A molecular dymamics simulation of a two-dimensional Lennard-Jones fluid.
#
# Author: david
#

using Printf
using StaticArrays

# Simulation parameters
const num_particles = 500
const force_cutoff = 3.0
const force_cutoff_2 = force_cutoff * force_cutoff
const p_eat_cutoff = 4.0 * ((force_cutoff^-12) - (force_cutoff^-6))
const target_temperature = 0.42
const time_step = 0.020
const dt_over_2 = 0.5 * time_step
const dt_squared_over_2 = 0.5 * time_step * time_step
const box_width = 100.0 # In units of molecular diameter.
const box_width_minus_half = box_width - 0.5
const box_height = 100.0
const box_height_minus_half = box_height - 0.5
const wall_stiffness = 50.0 # "Spring constant" for walls.
const equilibration_time = 200.0
const steps_between_equil_rescaling = 10
const production_time = 200.0
const steps_between_prod_rescaling = 100
const steps_per_printout = 50.0

const x = @MVector zeros(num_particles)
const y = @MVector zeros(num_particles)
const vx = @MVector zeros(num_particles)
const vy = @MVector zeros(num_particles)
const ax = @MVector zeros(num_particles)
const ay = @MVector zeros(num_particles)

const t = Ref{Float64}(0.0)
const steps_accomplished = Ref{Int64}(0)

const v_sum_x = Ref{Float64}(0.0)
const v_sum_y = Ref{Float64}(0.0)
const k_e = Ref{Float64}(0.0)
const p_e = Ref{Float64}(0.0)
const current_t = Ref{Float64}(0.0)
const total_t = Ref{Float64}(0.0)
const average_t = Ref{Float64}(0.0)
const pressure = Ref{Float64}(0.0)
const total_p = Ref{Float64}(0.0)
const average_p = Ref{Float64}(0.0)
const sample_count = Ref{Int64}(0)

function main()
    @printf "md2d - MD simulation of a 2D argon gas Lennard-Jones system.\n"
    @printf "This version is written in Julia and running version %s.\n" VERSION
    print_config()

    initialize_particles()

    @printf "\nEquilibration\n"
    print_properties_header()
    print_properties()
    start_ns = time_ns()
    while (t[] < equilibration_time)
        single_step()
        steps_accomplished[] += 1
        t[] += time_step
        if ((steps_accomplished[] % steps_per_printout) == 0)
            compute_properties()
            print_properties()
        end
        if ((steps_accomplished[] % steps_between_equil_rescaling) == 0)
            rescale_velocities()
        end
    end

    reset_measurements()
    rescale_velocities()
    @printf "\nProduction\n"
    print_properties_header()
    while (t[] < (equilibration_time + production_time))
        single_step()
        steps_accomplished[] += 1
        t[] += time_step
        if ((steps_accomplished[] % steps_per_printout) == 0)
            compute_properties()
            print_properties()
        end
        if ((steps_accomplished[] % steps_between_prod_rescaling) == 0)
            rescale_velocities()
        end
    end

    # Very crude calculation of steps per second. Includes print time.
    finish_ns = time_ns()
    elapsed_sec = (finish_ns - start_ns) / 1000000000
    @printf "Elapsed time: %.3f seconds, %.3f steps per second\n" elapsed_sec (
        steps_accomplished[] / elapsed_sec
    )
end

"Execute one time step using the Verlet algorithm."
function single_step()
    for i in eachindex(x)
        # Update positions.
        x[i] += (vx[i] * time_step) + (ax[i] * dt_squared_over_2)
        y[i] += (vy[i] * time_step) + (ay[i] * dt_squared_over_2)
        # Update velocities "halfway".
        vx[i] += (ax[i] * dt_over_2)
        vy[i] += (ay[i] * dt_over_2)
    end
    compute_accelerations()
    for i in eachindex(x)
        # Finish updating velocities using the new accelerations.
        vx[i] += (ax[i] * dt_over_2)
        vy[i] += (ay[i] * dt_over_2)
    end
end

"Compute accelerations of the molecules from their current positions using
 the Lennard-Jones potential."
function compute_accelerations()
    p_e[] = 0.0
    wall_force = 0.0

    # First, check for bouncing against the walls.
    for i in eachindex(x)
        if (x[i] < 0.5)
            ax[i] = wall_stiffness * (0.5 - x[i])
            wall_force += ax[i]
            p_e[] += 0.5 * wall_stiffness * (0.5 - x[i]) * (0.5 - x[i])
        elseif (x[i] > box_width_minus_half)
            ax[i] = wall_stiffness * (box_width_minus_half - x[i])
            wall_force -= ax[i]
            p_e[] +=
                0.5 *
                wall_stiffness *
                (box_width_minus_half - x[i]) *
                (box_width_minus_half - x[i])
        else
            ax[i] = 0.0
        end

        if (y[i] < 0.5)
            ay[i] = wall_stiffness * (0.5 - y[i])
            wall_force += ay[i]
            p_e[] = 0.5 * wall_stiffness * (0.5 - y[i]) * (0.5 - y[i])
        elseif (y[i] > box_height_minus_half)
            ay[i] = wall_stiffness * (box_height_minus_half - y[i])
            wall_force -= ay[i]
            p_e[] +=
                0.5 *
                wall_stiffness *
                (box_height_minus_half - y[i]) *
                (box_height_minus_half - y[i])
        else
            ay[i] = 0.0
        end
    end

    pressure[] = wall_force / (4.0 * box_width)

    # Next, compute interactions using the Lennard-Jones potential.
    for i in eachindex(x)
        for j = firstindex(x):(i-1)
            dx = x[i] - x[j]
            dx2 = dx * dx
            # Make sure the pair are close enough to bother.
            if (dx2 < force_cutoff_2)
                dy = y[i] - y[j]
                dy2 = dy * dy
                if (dy2 < force_cutoff_2)
                    r_squared = dx2 + dy2
                    if (r_squared < force_cutoff_2)
                        r_squared_inv = 1.0 / r_squared
                        attract = r_squared_inv * r_squared_inv * r_squared_inv
                        repel = attract * attract
                        p_e[] += (4.0 * (repel - attract)) - p_eat_cutoff
                        f_over_r = 24.0 * ((2.0 * repel) - attract) * r_squared_inv
                        fx = f_over_r * dx
                        fy = f_over_r * dy
                        # Add the force on to i's acceleration.
                        ax[i] += fx
                        ay[i] += fy
                        ax[j] -= fx # Newton's 3rd law
                        ay[j] -= fy
                    end
                end
            end
        end
    end
end

"Reset accumulators and counters used for some measurements."
function reset_measurements()
    total_t[] = 0.0
    total_p[] = 0.0
    sample_count[] = 0
end

"Return the instantaneous kinetic energy."
function calculate_kinetic_energy()
    ek = 0.0
    for i in eachindex(vx)
        ek += 0.5 * (vx[i] * vx[i] + vy[i] * vy[i])
    end
    return ek
end

"Compute accumulated property values from sampled values."
function compute_properties()
    sample_count[] += 1
    k_e[] = calculate_kinetic_energy()
    current_t[] = k_e[] / num_particles
    total_t[] += current_t[]
    average_t[] = total_t[] / sample_count[]
    total_p[] += pressure[]
    average_p[] = total_p[] / sample_count[]
end

const nudge_count = Ref{Int64}(0)

"Return a small random number uniformly distributed around 0
 with a maximum magnitude of epsilon."
function nudge(epsilon)
    return (algo_647_uniform() - 0.5) * epsilon
end

"Place the particles in empty locactions of a grid on the simulation
 box starting from the lower left. Applies a small amount of \"jitter\"
 to break up the regularity of the layout a little. Does not check for
 overflow of the grid (too many particles to fit)."
function place_particles()
    spacing = 1.3 # Minimum space between particle centers.
    half_spacing = spacing / 2.0
    bwmhs = box_width - half_spacing
    jitter = 0.01 # Random amount to break up regularity.
    x_pos = half_spacing
    y_pos = box_height - half_spacing

    for i in eachindex(x)
        x[i] = x_pos + nudge(jitter)
        y[i] = y_pos + nudge(jitter)
        x_pos += spacing
        if (x_pos > bwmhs)
            x_pos = half_spacing
            y_pos -= spacing
        end
    end
end

"Re-scale the velocities to the target temperature."
function rescale_velocities()
    velocity_squared_sum = 0.0
    for i in eachindex(vx)
        velocity_squared_sum += vx[i] * vx[i] + vy[i] * vy[i]
    end
    scaling_factor = 2.0 * num_particles * target_temperature / velocity_squared_sum
    scaling_factor = sqrt(scaling_factor)
    for i in eachindex(vx)
        vx[i] *= scaling_factor
        vy[i] *= scaling_factor
    end
end

"Remove any net momentum from the system of particles."
function remove_drift()
    for i in eachindex(x)
        v_sum_x[] += vx[i]
        v_sum_y[] += vy[i]
    end
    for i in eachindex(x)
        vx[i] -= v_sum_x[] / num_particles
        vy[i] -= v_sum_y[] / num_particles
    end
end

"Initialize particle positions, scale velocities to target temperature,
 and remove systematic drift."
function initialize_particles()
    for i in eachindex(x)
        x[i] = 0.0
        y[i] = 0.0
        vx[i] = gaussian_deviate_marsaglia() - 0.5
        vy[i] = gaussian_deviate_marsaglia() - 0.5
        ax[i] = 0.0
        ay[i] = 0.0
    end
    place_particles()
    rescale_velocities()
    remove_drift()
end

#
# Printing stuff.
#

function print_config()
    @printf "\n Simulation Configuration:\n"
    @printf "   Number of particles        : %d\n" num_particles
    @printf "   Cutoff radius              : %.2f\n" force_cutoff
    @printf "   Target temperature         : %.3f\n" target_temperature
    @printf "   Integration step size      : %.3f\n" time_step
    @printf "   Width of simulation area   : %.1f\n" box_width
    @printf "   Height of simulation area  : %.1f\n" box_height
    @printf "   Area of simulation         : %.1f\n" box_width * box_height
    @printf "   Wall stiffness             : %.1f\n" wall_stiffness
    @printf "   Equilibration time         : %.1f\n" equilibration_time
    @printf "      Steps between rescaling : %d\n" steps_between_equil_rescaling
    @printf "   Production time            : %.1f\n" production_time
    @printf "      Steps between rescaling : %d\n" steps_between_prod_rescaling
    @printf "   Steps per print out        : %d\n" steps_per_printout
end

function print_properties_header()
    @printf "Time      Temp.   Pressure Tot. E.   Kin. E.   Pot. E.   Steps\n"
end

function print_properties()
    total_e = k_e[] + p_e[]
    @printf "%7.3f,  %5.3f,  %6.4f,  %7.2f,  %6.2f,  %8.2f,  %d\n" t[] average_t[] average_p[] total_e k_e[] p_e[] steps_accomplished[]
end

"Print some pseudo random numbers. Just for debugging purposes."
function print_random_numbers()
    @printf "\nRandom numbers from a uniform distribution.\n"
    for i = 0:9
        @printf "Uniform #%i: %.17f\n" i algo_647_uniform()
    end
    @printf "\nRandom numbers from a normal distribution.\n"
    for i = 0:9
        @printf "Normal #%i: %.17f\n" i gaussian_deviate_marsaglia()
    end
end

#
# Random number stuff.
#

# The algo_647_uniform function returns a random uniform deviate in
# the range [0, 1) calculated using the ACM algorithm 647 as
# recorded at: http://calgo.acm.org.
# The algorithm is from the paper "ALGORITHM 647: Implementation and
# relative Efficiency of Quasirandom Sequence Generators" by Bennet L. Fox,
# ACM Transactions on Mathematical Software (TOMS), Volume 12, No 4,
# Dec 1986, pp. 362-376.
#
# I have never seen the actual paper - it's behind a paywall - but the
# FORTRAN version of the algorithm is easy enough to understand.

# Something to hold an mutable seed value for the random number generator.
mutable struct seedType
    seed::Int32
end

# Something to hold an mutable seed value for the random number generator.
glbl_seed = seedType(12345)

"Allow the seed to be set externally."
function algo_647_set_seed(new_seed)
    setproperty!(glbl_seed, :seed, new_seed)
end

"Return a random uniform deviate in the range 0 <= x < 1.0 as a Float64."
function algo_647_uniform()
    k = floor(glbl_seed.seed / 127773)
    partial = 16807 * (glbl_seed.seed - k * 127773) - k * 2836
    if (partial < 0)
        setproperty!(glbl_seed, :seed, partial + typemax(Int32))
    else
        setproperty!(glbl_seed, :seed, partial)
    end
    return glbl_seed.seed * 4.656612875e-10
end

# The following is an implementation of the Marsaglia polar
# method for generating random standard normal deviates. See
# https://en.wikipedia.org/wiki/Marsaglia_polar_method.

mutable struct spareType
    has_spare::Bool
    spare::Float64
end

glbl_spare = spareType(false, 0.0)

"Return a random standard normal deviate as a double."
function gaussian_deviate_marsaglia()
    fac = 0.0
    rsq = 0.0
    r1 = 0.0
    r2 = 0.0
    if (glbl_spare.has_spare)
        setproperty!(glbl_spare, :has_spare, false)
        return glbl_spare.spare
    else
        while true
            r1 = 2.0 * algo_647_uniform() - 1.0
            r2 = 2.0 * algo_647_uniform() - 1.0
            rsq = r1 * r1 + r2 * r2
            (rsq >= 1.0 || rsq == 0.0) || break
        end
        fac = sqrt(-2.0 * log(rsq) / rsq)
        setproperty!(glbl_spare, :spare, r1 * fac)
        setproperty!(glbl_spare, :has_spare, true)
        return r2 * fac
    end
end

main()
